ent/encryption/utils.cc

```diff
@@ -0,0 +1,151 @@
+#include <seastar/util/short_streams.hh>
+
+#include "utils.hh"
+
+using namespace seastar;
+
+encryption::httpclient::httpclient(std::string host, uint16_t port, seastar::shared_ptr<seastar::tls::certificate_credentials> creds)
+    : _host(std::move(host))
+    , _port(port)
+    , _creds(std::move(creds))
+    , _req(http::request::make(httpd::operation_type::GET, _host, ""))
+{
+    _req._version = "1.1";
+}
+
+encryption::httpclient& encryption::httpclient::add_header(std::string_view key, std::string_view value) {
+    _req._headers[sstring(key)] = sstring(value);
+    return *this;
+}
+
+void encryption::httpclient::clear_headers() {
+    _req._headers.clear();
+}
+
+seastar::future<encryption::httpclient::result_type> encryption::httpclient::send() {
+    result_type res;
+    co_await send([&](const http::reply& r, std::string_view body) {
+        res.reply._status = r._status;
+        res.reply._content = sstring(body);
+        res.reply._headers = r._headers;
+        res.reply._version = r._version;
+    });
+    co_return res;
+}
+
+/**
+ * Note: the client object is a one-off send-recive. We don't keepalive across calls or anything.
+ * This because 
+ * 1.) We generally never do more than one call every 20s or so to a given endpoint
+ * 2.) Credentials used (oath/AWS hash) are expiring, thus since we don't really do repeated 
+ *     calls anyway (only one to each endpoint every 20s), simply having a new auth token each
+ *     call simplifies state management.
+ * And keeps this code nice and simple.
+ */
+seastar::future<> encryption::httpclient::send(const handler_func& f) {
+    auto addr = co_await net::dns::resolve_name(_host, net::inet_address::family::INET /* TODO: our client does not handle ipv6 well?*/);
+
+    // NOTE: similar to utils::http::dns_connection_factory, but that type does
+    // not properly handle numeric hosts (don't validate certs for those)
+    class my_connection_factory : public http::experimental::connection_factory {
+        socket_address _addr;
+        shared_ptr<tls::certificate_credentials> _creds;
+        sstring _host;
+    public:
+        my_connection_factory(socket_address addr, shared_ptr<tls::certificate_credentials> creds, sstring host)
+            : _addr(std::move(addr))
+            , _creds(std::move(creds))
+            , _host(std::move(host))
+        {}
+        future<connected_socket> make(abort_source* as) override {
+            // don't verify host cert name if "host" is just an ip address.
+            // typically testing.
+            bool is_numeric_host = seastar::net::inet_address::parse_numerical(_host).has_value();
+            connected_socket s = co_await (_creds 
+                ? http::experimental::tls_connection_factory(_addr, _creds, is_numeric_host ? sstring{} : _host).make(as) 
+                : http::experimental::basic_connection_factory(_addr).make(as)
+            );
+            s.set_keepalive(true);
+            s.set_nodelay(true);
```

Comment:
This is **TCP**'s keepalive an nodelay? Why mess with them?

Reply:
The original code in the AWS SDK from which the auth/REST code was transposed (back ages ago) IIRC had these set. It has followed along. Probably not useful (at least not keepalive), but probably not hurtful either.